### ISSUE TITLE
bp_hist function gives undefined columns selected

### ISSUE THREAD

**Issue by chrisb2** [johnschwenck/bp#2]
bp installed with _devtools::install_github("johnschwenck/bp")_

I have import my data into data frame _bp_, and if I run:
```
bp_proc <- process_data(bp, sbp = 'SYS', dbp = 'DIA', hr = 'PUL', date_time = 'DT')
head(bp_proc)
```
The result is for _process_data_ is:
```
No PP column found. Automatically generated from SBP and DBP columns.
No MAP column found. Automatically generated from SBP and DBP columns.
NOTE: Created DATE column from DATE_TIME column
```
and _head_ gives:
```
        DATE           DATE_TIME HOUR SBP DBP BP_CLASS HR       MAP PP
1 2021-06-20 2021-06-20 09:53:00    9 114  75   Normal 61  88.00000 39
2 2021-06-20 2021-06-20 09:47:00    9 123  78 Elevated 61  93.00000 45
3 2021-06-18 2021-06-18 20:46:00   20 125  84 IDH - S1 64  97.66667 41
4 2021-06-18 2021-06-18 20:40:00   20 133  87  Stage 1 64 102.33333 46
5 2021-06-17 2021-06-17 19:17:00   19 116  81 IDH - S1 66  92.66667 35
6 2021-06-17 2021-06-17 19:11:00   19 132  85  Stage 1 62 100.66667 47
  SBP_CATEGORY DBP_CATEGORY ID GROUP TIME_OF_DAY DAY_OF_WEEK BP_TYPE
1       Normal       Normal  1     1     Morning         Sun    HBPM
2     Elevated       Normal  1     1     Morning         Sun    HBPM
3     Elevated      Stage 1  1     1     Evening         Fri    HBPM
4      Stage 1      Stage 1  1     1     Evening         Fri    HBPM
5       Normal      Stage 1  1     1     Evening         Thu    HBPM
6      Stage 1      Stage 1  1     1     Evening         Thu    HBPM
```
The names function gives:
```
> names(bp_proc)
 [1] "DATE"         "DATE_TIME"    "HOUR"         "SBP"          "DBP"
 [6] "BP_CLASS"     "HR"           "MAP"          "PP"           "SBP_CATEGORY"
[11] "DBP_CATEGORY" "ID"           "GROUP"        "TIME_OF_DAY"  "DAY_OF_WEEK"
[16] "BP_TYPE"
```
If I run _bp_hist(bp_proc)_ I get:
```
Error in `[.data.frame`(SBP_ToD, , c(3, 1, 2, 4, 5)) :
  undefined columns selected
```

This also happens with the version installed with _install.packages("bp")_.

What am I doing wrong?

thanks,
Chris

**Comment by johnschwenck** [comment]
Hi Chris,Thanks for bringing that to my attention. I'm a2ay from my computer at the moment, but I will investigate the problem and get back to you by the end of the week. Apologies for the setback.Best,John Schwenck
-------- Original message --------From: Chris Borrill ***@***.***> Date: 6/19/21  11:21 PM  (GMT-06:00) To: johnschwenck/bp ***@***.***> Cc: Subscribed ***@***.***> Subject: [johnschwenck/bp] bp_hist function gives undefined columns selected (#2) 
bp installed with devtools::install_github("johnschwenck/bp")
I have import my data into data frame bp, and if I run:
bp_proc <- process_data(bp, sbp = 'SYS', dbp = 'DIA', hr = 'PUL', date_time = 'DT')
head(bp_proc)

The result is for process_data is:
No PP column found. Automatically generated from SBP and DBP columns.
No MAP column found. Automatically generated from SBP and DBP columns.
NOTE: Created DATE column from DATE_TIME column

and head gives:
        DATE           DATE_TIME HOUR SBP DBP BP_CLASS HR       MAP PP
1 2021-06-20 2021-06-20 09:53:00    9 114  75   Normal 61  88.00000 39
2 2021-06-20 2021-06-20 09:47:00    9 123  78 Elevated 61  93.00000 45
3 2021-06-18 2021-06-18 20:46:00   20 125  84 IDH - S1 64  97.66667 41
4 2021-06-18 2021-06-18 20:40:00   20 133  87  Stage 1 64 102.33333 46
5 2021-06-17 2021-06-17 19:17:00   19 116  81 IDH - S1 66  92.66667 35
6 2021-06-17 2021-06-17 19:11:00   19 132  85  Stage 1 62 100.66667 47
  SBP_CATEGORY DBP_CATEGORY ID GROUP TIME_OF_DAY DAY_OF_WEEK BP_TYPE
1       Normal       Normal  1     1     Morning         Sun    HBPM
2     Elevated       Normal  1     1     Morning         Sun    HBPM
3     Elevated      Stage 1  1     1     Evening         Fri    HBPM
4      Stage 1      Stage 1  1     1     Evening         Fri    HBPM
5       Normal      Stage 1  1     1     Evening         Thu    HBPM
6      Stage 1      Stage 1  1     1     Evening         Thu    HBPM

If I run bp_hist(bp_proc) I get:
Error in `[.data.frame`(SBP_ToD, , c(3, 1, 2, 4, 5)) :
  undefined columns selected

This also happens with the version installed with install.packages("bp").
What am I doing wrong?
thanks,
Chris

**Comment by johnschwenck** [comment]
Hi Chris,

I just want to give you an update that I have passed this to my
collaborator (Irina Gaynanova) who has agreed to investigate the issue. We
hope to have a solution for you shortly. Dr. Gaynanova has been CC'ed to
this email for future correspondence.

Best,
John

**Comment by irinagain** [comment]
Hi Chris, I looked into the issue, and it looks like process_data worked well (the messages are just alerts that some columns are created automatically, they are not errors), but the bp_tables gives errors on a subsetting. I have tried on our example datasets, and I can reproduce the errors when we subset the data. I think it has to do with not having all the categories present so we will work on a patch in the meantime, and update you once it is done. Thank you for alerting us to the issue!

**Comment by chrisb2** [comment]
Thanks for your help. Curiously I cannot reproduce this using latest source from github using:
```
library(bp)
data("bp_jhs")
jhs_proc <- process_data(bp_jhs, sbp = 'sys.mmhg.', dbp = 'dias.mmhg.', date_time = 'datetime', hr = 'PULSE.BPM.')
head(bp_mag(jhs_proc))
tail(bp_mag(jhs_proc, inc_date = TRUE))
bp_hist(jhs_proc)
```
How did you reproduce using your data?

Chris

**Comment by irinagain** [comment]
Hi Chis, I just tried the same code with what is on GitHub right now and it works on my machine (MAC OS). Note that if you installed it earlier from GitHub, you may need to do the following to update to the most recent version:
`devtools::install_github("johnschwenck/bp", force = TRUE)`
In case it still doesn't work, can you please post the errors you are getting to help us debug? Thank you!

**Comment by chrisb2** [comment]
Successfully update (I select All when asked what to update). 

My test program is:
```
library(bp)

setwd("C:/aProject/blood-pressure-r")
bp <- read.csv(file = 'BloodPressure_20-06-2021.csv')
bp$DT <- paste(bp$DATE, bp$TIME)
bp <- within(bp, rm('BPZ', 'TAGS', 'DATE', 'TIME'))
bp_proc <- process_data(bp, sbp = 'SYS', dbp = 'DIA', hr = 'PUL', date_time = 'DT', dt_fmt='d/m/Y H:M')
bp_hist(bp_proc)
```
Using traceback, bp_hist function call gives:
```
> options(error = function() traceback(3))
> bp_hist(bp_proc)
Error in `[.data.frame`(SBP_ToD, , c(3, 1, 2, 4, 5)) :
  undefined columns selected
5: stop("undefined columns selected")
4: `[.data.frame`(SBP_ToD, , c(3, 1, 2, 4, 5))
3: SBP_ToD[, c(3, 1, 2, 4, 5)]
2: bp_tables(data)
1: bp_hist(bp_proc)
```
names function gives:
```
> names(bp_proc)
 [1] "DATE"         "DATE_TIME"    "HOUR"         "SBP"          "DBP"
 [6] "BP_CLASS"     "HR"           "MAP"          "PP"           "SBP_CATEGORY"
[11] "DBP_CATEGORY" "ID"           "GROUP"        "TIME_OF_DAY"  "DAY_OF_WEEK"
[16] "BP_TYPE"
```
View(bp_proc) shows all cols appear to be populated with no missing values or NAs.

Here is 10 rows of my CSV data file which replicate the error:
```
DATE,TIME,SYS,DIA,PUL,BPZ,TAGS
20/06/2021,09:53,114,75,61,Normal,Left arm | Seated
20/06/2021,09:47,123,78,61,Prehypertension,Left arm
18/06/2021,20:46,125,84,64,Prehypertension,Left arm | Seated
18/06/2021,20:40,133,87,64,Prehypertension,Left arm | Seated
17/06/2021,19:17,116,81,66,Prehypertension,Left arm | Seated
17/06/2021,19:11,132,85,62,Prehypertension,Left arm | Seated
15/06/2021,19:48,124,79,64,Prehypertension,Left arm | Seated
15/06/2021,19:43,126,82,63,Prehypertension,Left arm | Seated
14/06/2021,19:18,126,87,60,Prehypertension,Left arm | Seated
14/06/2021,19:13,135,85,59,Prehypertension,Left arm | Seated
```
thanks for your help,
Chris

**Comment by irinagain** [comment]
Hi Chris, thank you, to clarify, we have not yet fixed the bug that appeared on your data yet (but we hope soon). What I meant that the code should work without any issues on the provided example data (e.g. bp_jhs as in the code you submitted). The reason there is a bug on your data is that some categories are not there, which requires rewrite on our part. We are working on the fix but it will take some time. I hope we have it soon, and we will update your here once we update Github. Irina

**Comment by irinagain** [comment]
Hi Chris, I just pushed a fix that I believe should make bp_tables and bp_hist work as expected on your data. If you can reinstall from Github 
`devtools::install_github("johnschwenck/bp", force = TRUE)`
and let us know whether you still have issues or not, we will appreciate it. Thank you for your patience!

**Comment by chrisb2** [comment]
`bp_hist` , `bp_report`, `bp_scatter` and `dow_tod_plots` are working now. Thanks very much.

Have notice a few issues with graphics in these reports, but will raise a separate issue for those.

Chris
